File: prop_clingo/propagator_clingo_py/propagator_clingo_initializer.py
from typing import Dict, List, Optional
from clingo import Symbol, PropagateInit, SymbolicAtoms
import clingo

class PropagatorClingoInitializer:
    _instance: Optional["PropagatorClingoInitializer"] = None

    def __init__(self):
        if PropagatorClingoInitializer._instance is not None:
            raise Exception("This class is a singleton!")
        self.first = True
        self.atomNames: Optional[Dict[str, int]] = None
        self.map_slit_plit: Optional[Dict[int, List[int]]] = None
        self.map_plit_slit: Optional[Dict[int, int]] = None
        self.nt = 0
        self.max_plit = 0
        self.lits: Optional[List[int]] = None
        PropagatorClingoInitializer._instance = self

    @staticmethod
    def get_instance():
        if PropagatorClingoInitializer._instance is None:
            PropagatorClingoInitializer._instance = PropagatorClingoInitializer()
        return PropagatorClingoInitializer._instance

    def init(self, _init: clingo.PropagateInit, propagator):
        if not self.first:
            propagator.atomNames = self.atomNames
            propagator.map_slit_plit = self.map_slit_plit
            propagator.map_plit_slit = self.map_plit_slit
            return
        
        self.first = False

        self.atomNames = {}
        self.map_slit_plit = {}
        self.map_plit_slit = {}
        self.nt = _init.number_of_threads
        print(f"[init] number of threads: {self.nt}")

        symbolic_atoms: SymbolicAtoms = _init.symbolic_atoms
        for atom in symbolic_atoms:
            symbol = atom.symbol
            plit = atom.literal
            slit = _init.solver_literal(plit)

            if plit > self.max_plit:
                self.max_plit = plit
            
            self.atomNames[symbol] = plit
            self.map_plit_slit[plit] = slit
            self.map_plit_slit[-plit] = -slit

            if slit not in self.map_slit_plit:
                self.map_slit_plit[slit] = []
            self.map_slit_plit[slit].append(plit)

            if -slit not in self.map_slit_plit:
                self.map_slit_plit[-slit] = []
            self.map_slit_plit[-slit].append(-plit)

        self.lits = [self.max_plit]

        facts = self.map_slit_plit.get(1, [])
        self.lits.extend(facts)

        propagator.atomNames = self.atomNames
        propagator.map_slit_plit = self.map_slit_plit
        propagator.map_plit_slit = self.map_plit_slit

    @staticmethod
    def cleanup():
        PropagatorClingoInitializer._instance = None

File: amosum/ijcai_version/AMOSUM/prop_clingo/propagator_clingo_c/propagator_clingo_initializer.cpp
#pragma once
#include "propagator_clingo_initializer.h"


void PropagatorClingoInitializer::init(clingo_propagate_init* _init, PropagatorClingo& propagator){

    if(!first){
        propagator.atomNames = atomNames;
        propagator.map_slit_plit = map_slit_plit.get();
        propagator.map_plit_slit = map_plit_slit.get();
        return ;
    }
    first = false ;

    atomNames = new std::unordered_map<clingo_symbol_t, clingo_literal_t>();
 
    map_slit_plit.reset(new std::unordered_map<clingo_literal_t, std::vector<clingo_literal_t>>());
    map_plit_slit.reset(new std::unordered_map<clingo_literal_t, clingo_literal_t>());
    nt = clingo_propagate_init_number_of_threads(_init);

    debug("[init] number of threads ", nt);
    clingo_symbolic_atoms_t const *symbolic_atoms;
    clingo_symbolic_atom_iterator_t symbolic_atoms_it, symbolic_atoms_ie;
    
    handle_error(clingo_propagate_init_symbolic_atoms(_init, &symbolic_atoms));
    handle_error(clingo_symbolic_atoms_end(symbolic_atoms, &symbolic_atoms_ie));

    handle_error(clingo_symbolic_atoms_begin(symbolic_atoms, NULL, &symbolic_atoms_it));

    while (true) {
        bool equal;
        clingo_literal_t plit;
        clingo_literal_t slit;
        clingo_symbol_t symbol;

        // stop iteration if the end is reached
        handle_error(clingo_symbolic_atoms_iterator_is_equal_to(symbolic_atoms, symbolic_atoms_it, symbolic_atoms_ie, &equal));
        if (equal) { break; }
        handle_error(clingo_symbolic_atoms_symbol(symbolic_atoms, symbolic_atoms_it, &symbol));
        std::string symbol_str = from_symbol_to_string(symbol) ;

        
        handle_error(clingo_symbolic_atoms_literal(symbolic_atoms, symbolic_atoms_it, &plit));
        handle_error(clingo_propagate_init_solver_literal(_init, plit, &slit));

        if (plit > max_plit) { max_plit = plit; }

        // debugf("[init] symbol: ", symbol_str, " symbol: ", symbol, " plit: ", plit, " slit: ", slit)

        atomNames->emplace(symbol, plit);
        (*map_plit_slit)[plit] = slit ; 
        (*map_plit_slit)[not_(plit)] = not_(slit) ; 
        update_map_value_vector<clingo_literal_t, clingo_literal_t>(*map_slit_plit, slit, plit);
        update_map_value_vector<clingo_literal_t, clingo_literal_t>(*map_slit_plit, not_(slit), not_(plit));
    
        clingo_symbolic_atoms_next(symbolic_atoms, symbolic_atoms_it, &symbolic_atoms_it);
    }
    
    lits = new std::vector<clingo_literal_t>{max_plit};

    std::vector<clingo_literal_t> facts = get_map_value_vector<clingo_literal_t, clingo_literal_t>((*map_slit_plit), 1);
    extend_vector(*lits, facts);

    propagator.atomNames = atomNames;
    propagator.map_slit_plit = map_slit_plit.get();
    propagator.map_plit_slit = map_plit_slit.get();
}

PropagatorClingoInitializer* PropagatorClingoInitializer::instance = nullptr;